replica/table.cc

```diff
@@ -2059,40 +2056,45 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
+        for (auto& [id, sg] : storage_groups) {
+            if (!sg->main_compaction_group()->empty()) {
+                on_internal_error(tlogger, format("Found that storage of group {} for table {} wasn't split correctly, " \
+                                                "therefore groups cannot be remapped with the new tablet count.",
+                                                id, table_id));
+            }
+            // Remove old main groups, they're unused, but they need to be deregistered properly
+            auto cg_ptr = std::move(sg->main_compaction_group());
+            auto f = cg_ptr->stop();
+            if (!f.available() || f.failed()) [[unlikely]] {
+                stop_fut = stop_fut.then([f = std::move(f), cg_ptr = std::move(cg_ptr)] () mutable {
+                    return std::move(f).handle_exception([cg_ptr = std::move(cg_ptr)] (std::exception_ptr ex) {
+                        tlogger.warn("Failed to stop compaction group: {}.  Ignored", std::move(ex));
+                    });
                 });
-            });
-        }
-        unsigned first_new_id = id << growth_factor;
-        auto split_ready_groups = std::move(*sg).split_ready_compaction_groups();
-        if (split_ready_groups.size() != split_size) {
-            on_internal_error(tlogger, format("Found {} split ready compaction groups, but expected {} instead.", split_ready_groups.size(), split_size));
-        }
-        for (unsigned i = 0; i < split_size; i++) {
-            auto group_id = first_new_id + i;
-            split_ready_groups[i]->update_id_and_range(group_id, new_tmap.get_token_range(locator::tablet_id(group_id)));
-            new_storage_groups[group_id] = std::make_unique<storage_group>(std::move(split_ready_groups[i]), nullptr);
-        }
+            }
+            unsigned first_new_id = id << growth_factor;
+            auto split_ready_groups = std::move(*sg).split_ready_compaction_groups();
+            if (split_ready_groups.size() != split_size) {
+                on_internal_error(tlogger, format("Found {} split ready compaction groups, but expected {} instead.", split_ready_groups.size(), split_size));
+            }
+            for (unsigned i = 0; i < split_size; i++) {
+                auto group_id = first_new_id + i;
+                split_ready_groups[i]->update_id_and_range(group_id, new_tmap.get_token_range(locator::tablet_id(group_id)));
+                new_storage_groups[group_id] = std::make_unique<storage_group>(std::move(split_ready_groups[i]), nullptr);
+            }
 
-        tlogger.debug("Remapping tablet {} of table {} into new tablets [{}].",
-                      id, table_id, fmt::join(boost::irange(first_new_id, first_new_id+split_size), ", "));
-    }
+            tlogger.debug("Remapping tablet {} of table {} into new tablets [{}].",
+                        id, table_id, fmt::join(boost::irange(first_new_id, first_new_id+split_size), ", "));
+        }
+    });
```

Comment:
we need to run entire split completion with exclusive lock, not only the map update. that's because it will already be moving groups around and that can cause a problem with concurrent scanning. it could also prevent a theoretical race between multiple split completion handling, although I find that very hard (maybe impossible?) to happen in practice.

Reply:
please document it.